rust_src/src/hashtable.rs

```diff
@@ -327,4 +401,126 @@ fn define_hash_table_test(name: LispObject, test: LispObject, hash: LispObject)
+    let mut table = LispHashTableRef::new(h);
+    let n = unsafe { gc_asize(table.index) };
+    let mut marked = false;
+
+    for bucket in 0..n {
+        let mut prev = -1;
+        let mut next;
+        let mut i = table.get_index_slot(bucket);
+        while 0 <= i {
+            let key_survives = unsafe { survives_gc_p(table.get_hash_key(i).to_raw()) };
+            let value_survives = unsafe { survives_gc_p(table.get_hash_value(i).to_raw()) };
+            let remove_p = match table.weak {
+                Qkey => !key_survives,
+                Qvalue => !value_survives,
+                Qkey_or_value => !(key_survives || value_survives),
+                Qkey_and_value => !(key_survives && value_survives),
+                _ => panic!(),
+            };
+
+            next = table.get_next_slot(i);
+
+            if remove_entries_p {
+                if remove_p {
+                    if prev < 0 {
+                        table.set_index_slot(bucket, next);
+                    } else {
+                        table.set_next_slot(prev, next);
+                    }
+
+                    table.set_next_slot(i, table.next_free);
+                    table.next_free = i;
+
+                    table.set_hash_key(i, LispObject::constant_nil());
+                    table.set_hash_value(i, LispObject::constant_nil());
+                    table.set_hash_hash(i, LispObject::constant_nil());
+
+                    table.count -= 1;
+                } else {
+                    prev = i;
+                }
+            } else {
+                if !remove_p {
+                    if !key_survives {
+                        unsafe { mark_object(table.get_hash_key(i).to_raw()) };
+                        marked = true;
+                    }
+
+                    if !value_survives {
+                        unsafe { mark_object(table.get_hash_value(i).to_raw()) };
+                        marked = true;
+                    }
+                }
+            }
+
+            i = next;
+        }
+    }
+
+    marked
+}
+
+/// Remove elements from weak hash tables that don't survive the
+/// current garbage collection.  Remove weak tables that don't survive
+/// from Vweak_hash_tables.  Called from gc_sweep.
+#[no_mangle]
+pub extern "C" fn sweep_weak_hash_tables() {
+    // Mark all keys and values that are in use.  Keep on marking until
+    // there is no more change.  This is necessary for cases like
+    // value-weak table A containing an entry X -> Y, where Y is used in a
+    // key-weak table B, Z -> Y.  If B comes after A in the list of weak
+    // tables, X -> Y might be removed from A, although when looking at B
+    // one finds that it shouldn't.
+
+    // This may look odd, but this is 'abusing' a Rust while loop to make
+    // a 'do-while' loop. All the logic is done in the conditional of the while loop,
+    // and the body is just empty.
+    let mut marked;
```

Comment:
Yuck. Why?

Reply:
Haha I think I was trying to get a little to clever here. Once I saw this worked I jumped the gun a little. I can update this to be more idiomatic Rust.

Reply:
At the very least it doesn't seem like you need the `let` outside the loop.

Reply:
We all have those moments.

A wise manager once taught me "Debugging is harder than coding. If you code to the edge of your skill by definition you cannot debug your own code". Save the cute code for when it really matters was an important lesson.